sycl/doc/extensions/proposed/sycl_ext_oneapi_load_store.asciidoc

```diff
@@ -0,0 +1,341 @@
+
+struct nontemporal { /* unspecified */ }; // 1
+struct temporal { /* unspecified */ }; // 2
+
+} // namespace experimental
+} // namespace extg
+} // namespace oneapi
+} // namespace sycl
+```
+
+1:: indicates that the associated memory be accessed in as maximally nontemporal
+a fashion as possible.
+2:: indicates that the associated memory be accessed in
+as maximally temporal a fashion as possible.
+
+=== Properties
+
+The nontemporal properties that are parameterized by the above are:
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+struct temporality_hint_key {
+  template <typename T>
+  using value_t = property_value<temporality_hint_key, T>;
+};
+
+struct L1_cache_hint_key {
+  template <typename T>
+  using value_t = property_value<L1_cache_hint_key, T>;
+};
+
+struct L2_cache_hint_key {
+  template <typename T>
+  using value_t = property_value<L2_cache_hint_key, T>;
+};
+
+struct L3_cache_hint_key  {
+  template <typename T>
+  using value_t = property_value<L3_cache_hint_key, T>;
+};
+
+struct L4_cache_hint_key {
+  template <typename T>
+  using value_t = property_value<L4_cache_hint_key, T>;
+};
+
+} // namespace experimental
+} // namespace ext
+} // namespace oneapi
+} // namespace sycl
+```
+
+The `temporality_hint_key` property is the most generic and it should override
+any other nontemporal properties, if present.
+
+The property values as passed to the `{L1,L2,L3,L4}_cache_hint_key` property
+classes should apply only to the cache level specified; the precise mapping to
+hardware constructs is otherwise implementation-defined.
+
+=== Convenience variables
+
+The following convenience variables help use the above when declaring property lists:
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+template <typename T>
+inline constexpr temporality_hint_key::value_t<T> temporality_hint;
+
+template <typename T>
+inline constexpr L1_cache_hint_key::value_t<T> L1_cache_hint;
+
+template <typename T>
+inline constexpr L2_cache_hint_key::value_t<T> L2_cache_hint;
+
+template <typename T>
+inline constexpr L3_cache_hint_key::value_t<T> L3_cache_hint;
+
+template <typename T>
+inline constexpr L4_cache_hint_key::value_t<T> L4_cache_hint;
+
+} // namespace experimental
+} // namespace ext
+} // namespace oneapi
+} // namespace sycl
+```
+
+== Implementation Notes
+
+These properties are intended to be hints to guide the compiler; specific
+nontemporal behavior should not be assumed.
+
+Most extant architectures lack awareness of categories of memory as they are
+understood by the programmer (i.e. buffers, arrays, structures) and only expose
+temporality controls at the granularity of memory-transacting instructions. This
+extension provides a groundwork for future extensions that expose pointer- and
+accessor-level semantics. A future extension may provide more
+architecture-specific hints and coarser controls for applying hints.
```

Comment:
(adding a global comment here only so it can be a threaded conversation)

After reviewing #5755, I'm wondering if we can use `annotated_ptr` to convey the properties, and then change this extension to use `annotated_ptr`.  If we did this:

* The work-item granularity functions (`load()` and `store()`) would be unnecessary because we would just use `annotated_ptr`:

  ```
  annotated_ptr aptr(ptr, properties{temporality_hint<nontemporal>});
  /*...*/ = *aptr;
  *aptr = 42;
  ```

* The joint group functions would use either `annotated_ptr` or a raw pointer.  The `annotated_ptr` usage would be like this:

  ```
  // In convergent code
  /*...*/ = joint_load(g, aptr);
  ```

* The group block functions would also use either `annotated_ptr` or a raw pointer.  The `annotated_ptr` usage would be like this:

  ```
  // In convergent code
  group_block_store(g, aptr, val);
  ```

Reply:
Oh, man. I guess you don't remember that `annotated_ptr` work came out of the original proposal for this, and we chose to split this proposal into a 'basic' load/store version and a version that supports `annotated_ptr`. 

There are a few reasons for this: 

1. was that load/store was deemed simpler and easier to propose, and that `annotated_ptr` could build off of it. 
2. was that having a fine-grained control (load/store) is complementary to a coarse-grained control. load/store can override properties in an `annotated_ptr`, for example. 

Does that make sense? I think it absolutely makes sense to have this work with `annotated_ptr`, but I don't think we should drop any of the load/store functions.

Reply:
@gmlueck , did you see this?

Reply:
Thanks for the ping.  At this point we already have a proposed extension for `annotated_ptr`, so reason (1) is no longer a strong argument.  Do we think that point (2) (fine grained control) is enough justification to maintaining two ways to do non-temporal memory accesses?

Regardless, we should decide now how the non-temporary properties interact with `annotated_ptr`.

If we decided to use `annotated_ptr` **instead** of the load/store extension to do non-temporal accesses, I think we would just move the properties to that extension, and add overloads for `annotated_ptr` as I outline above.

If we keep the non-temporal support in the load/store extension, how would the interaction with `annotated_ptr` work?

Reply:
I do think (2) warranted. I see these as complimentary features. annotated_ptr + nontemporal properties allows for a productive, coarse-grained way of marking some semantically important piece of memory, while load/store allow for fine-grained expressions.

In my original proposal, I suggested that load/store would override any annotated properties, and I stand by that. It's a 'most specific' type of behavior that I think users can easily reason about. 

I should add that this was renamed "load_store" from "nontemporal" at the suggestion of @Pennycook; we are considering how to add marray/vec behavior as well.